File: files.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/rwcarlsen/goexif/exif"
)

type LocalFileInfo struct {
	Filename string
	Size     int64
	Date     time.Time
	Width    int
	Height   int
}

type RemoteItemInfo struct {
	ID           string
	CreationTime string
	Width        string
	Height       string
}

func FindMediaFiles(folder string) ([]string, error) {
	entries, err := os.ReadDir(folder)
	if err != nil {
		return nil, err
	}

	var files []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		ext := strings.ToLower(filepath.Ext(e.Name()))
		if SupportedExtensions[ext] {
			files = append(files, filepath.Join(folder, e.Name()))
		}
	}
	sort.Strings(files)
	return files, nil
}

func FormatSize(sizeBytes int64) string {
	switch {
	case sizeBytes < 1024:
		return fmt.Sprintf("%d B", sizeBytes)
	case sizeBytes < 1024*1024:
		return fmt.Sprintf("%.1f KB", float64(sizeBytes)/1024)
	case sizeBytes < 1024*1024*1024:
		return fmt.Sprintf("%.1f MB", float64(sizeBytes)/(1024*1024))
	default:
		return fmt.Sprintf("%.2f GB", float64(sizeBytes)/(1024*1024*1024))
	}
}

func GetLocalFileInfo(fpath string) LocalFileInfo {
	info := LocalFileInfo{
		Filename: filepath.Base(fpath),
	}

	stat, err := os.Stat(fpath)
	if err != nil {
		return info
	}
	info.Size = stat.Size()
	info.Date = stat.ModTime()

	f, err := os.Open(fpath)
	if err != nil {
		return info
	}
	defer f.Close()

	x, err := exif.Decode(f)
	if err == nil {
		if dt, err := x.DateTime(); err == nil {
			info.Date = dt
		}
		if w, err := x.Get(exif.PixelXDimension); err == nil {
			if v, err := w.Int(0); err == nil {
				info.Width = v
			}
		}
		if h, err := x.Get(exif.PixelYDimension); err == nil {
			if v, err := h.Int(0); err == nil {
				info.Height = v
			}
		}
	}

	return info
}

func FormatRemoteDate(creationTime string) string {
	if creationTime == "" {
		return "?"
	}
	s := strings.Replace(creationTime, "Z", "+00:00", 1)
	t, err := time.Parse(time.RFC3339, s)
	if err != nil {
		if len(creationTime) >= 16 {
			return creationTime[:16]
		}
		return creationTime
	}
	return t.Format("2006-01-02 15:04")
}

func PromptDuplicate(fpath string, remote RemoteItemInfo, reader *bufio.Reader) string {
	local := GetLocalFileInfo(fpath)
	filename := filepath.Base(fpath)

	localDate := local.Date.Format("2006-01-02 15:04")
	localSize := fmt.Sprintf("%8s", FormatSize(local.Size))
	localDim := ""
	if local.Width > 0 && local.Height > 0 {
		localDim = fmt.Sprintf("%d×%d", local.Width, local.Height)
	}

	remoteDate := FormatRemoteDate(remote.CreationTime)
	remoteDim := ""
	if remote.Width != "" && remote.Height != "" {
		remoteDim = fmt.Sprintf("%s×%s", remote.Width, remote.Height)
	}

	fmt.Printf("\n  ⚠️  Дубликат найден: %s\n", filename)
	fmt.Printf("       Локальный:  %s  |  %s  |  %s\n", localDate, localSize, localDim)
	fmt.Printf("       Удалённый:  %s  |  %8s  |  %s\n", remoteDate, "—", remoteDim)

	for {
		fmt.Print("       [S]kip / [R]eplace / Re[n]ame? ")
		line, _ := reader.ReadString('\n')
		choice := strings.TrimSpace(strings.ToLower(line))
		switch choice {
		case "s", "r", "n":
			return choice
		default:
			fmt.Println("       Введите s, r или n")
		}
	}
}

File: gphoto_export/files.py
from datetime import datetime
from pathlib import Path

from PIL import Image, ExifTags

from config import SUPPORTED_EXTENSIONS


def find_media_files(folder: Path) -> list[Path]:
    """Находит все поддерживаемые медиафайлы в папке (без рекурсии)."""
    files = []
    for f in sorted(folder.iterdir()):
        if f.is_file() and f.suffix.lower() in SUPPORTED_EXTENSIONS:
            files.append(f)
    return files


def format_size(size_bytes: int) -> str:
    """Форматирует размер в человекочитаемый вид."""
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f} KB"
    elif size_bytes < 1024 * 1024 * 1024:
        return f"{size_bytes / (1024 * 1024):.1f} MB"
    else:
        return f"{size_bytes / (1024 * 1024 * 1024):.2f} GB"


def get_local_file_info(filepath: Path) -> dict:
    """Возвращает информацию о локальном файле: размер, дату, размеры изображения."""
    stat = filepath.stat()
    info = {
        "filename": filepath.name,
        "size": stat.st_size,
        "date": None,
        "width": None,
        "height": None,
    }

    # Пробуем получить EXIF DateTimeOriginal
    try:
        with Image.open(filepath) as img:
            info["width"], info["height"] = img.size
            exif_data = img._getexif()
            if exif_data:
                for tag_id, value in exif_data.items():
                    tag_name = ExifTags.TAGS.get(tag_id, "")
                    if tag_name == "DateTimeOriginal":
                        info["date"] = datetime.strptime(value, "%Y:%m:%d %H:%M:%S")
                        break
    except Exception:
        pass

    if info["date"] is None:
        info["date"] = datetime.fromtimestamp(stat.st_mtime)

    return info


def format_remote_date(creation_time: str) -> str:
    """Форматирует дату из Google Photos API (ISO 8601)."""
    if not creation_time:
        return "?"
    try:
        dt = datetime.fromisoformat(creation_time.replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d %H:%M")
    except Exception:
        return creation_time[:16]


def prompt_duplicate(filepath: Path, remote_info: dict) -> str:
    """
    Показывает сравнение локального и удалённого файла,
    запрашивает действие: Skip / Replace / Rename.
    Возвращает 's', 'r' или 'n'.
    """
    local = get_local_file_info(filepath)

    local_date = local["date"].strftime("%Y-%m-%d %H:%M") if local["date"] else "?"
    local_dims = (f"{local['width']}×{local['height']}"
                  if local["width"] and local["height"] else "?")

    remote_date = format_remote_date(remote_info.get("creationTime", ""))
    rw, rh = remote_info.get("width", ""), remote_info.get("height", "")
    remote_dims = f"{rw}×{rh}" if rw and rh else "?"

    print(f"\n  ⚠️  Дубликат найден: {filepath.name}")
    print(f"       Локальный:  {local_date}  |  {format_size(local['size']):>8}  |  {local_dims}")
    print(f"       Удалённый:  {remote_date}  |  {'—':>8}  |  {remote_dims}")

    while True:
        choice = input("       [S]kip / [R]eplace / Re[n]ame? ").strip().lower()
        if choice in ("s", "r", "n"):
            return choice
        print("       Введите s, r или n")
